File: server/middleware/authentication.js
const { verifyToken } = require("../helpers/jsonwebtoken");
const { User } = require("../models");

const Authentication = async (req, res, next) => {
  try {
    // Mengambil header authorization
    const authHeader = req.headers["authorization"];

    // Memastikan header authorization ada
    if (!authHeader) {
      throw { name: "Unauthorized", message: "Authorization header missing" };
    }

    // Memisahkan token dari header
    const token = authHeader.split(" ")[1];
    if (!token) {
      throw { name: "Unauthorized", message: "Token missing" };
    }

    // Memverifikasi token
    const decoded = verifyToken(token);
    const user = await User.findByPk(decoded.id);

    // Memastikan user ada
    if (!user) {
      throw { name: "Unauthorized", message: "User not found" };
    }

    // Menyimpan informasi user di request
    req.user = user;
    next(); // Melanjutkan ke middleware berikutnya
  } catch (error) {
    console.log(error);
    next(error); // Meneruskan error ke middleware penanganan error
  }
};

module.exports = Authentication;

File: GOServer/middleware/authentication.go
package middlewares

import (
	"errors"
	"github.com/putram11/GudangKU/GOServer/models"
	"github.com/putram11/GudangKU/GOServer/helpers"
	"net/http"
	"strings"

	"github.com/labstack/echo/v4"
	"gorm.io/gorm"
)

// Authentication middleware
func Authentication(db *gorm.DB) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			// Mengambil header Authorization
			authHeader := c.Request().Header.Get("Authorization")
			if authHeader == "" {
				return echo.NewHTTPError(http.StatusUnauthorized, "Authorization header missing")
			}

			// Memisahkan token dari header
			tokenParts := strings.Split(authHeader, " ")
			if len(tokenParts) != 2 || tokenParts[0] != "Bearer" {
				return echo.NewHTTPError(http.StatusUnauthorized, "Invalid authorization format")
			}
			token := tokenParts[1]

			// Memverifikasi token
			decoded, err := helpers.VerifyToken(token)
			if err != nil {
				return echo.NewHTTPError(http.StatusUnauthorized, "Invalid or expired token")
			}

			// Memastikan user ada di database
			var user models.User
			if err := db.First(&user, "id = ?", decoded.ID).Error; err != nil {
				if errors.Is(err, gorm.ErrRecordNotFound) {
					return echo.NewHTTPError(http.StatusUnauthorized, "User not found")
				}
				return echo.NewHTTPError(http.StatusInternalServerError, "Database error")
			}

			// Menyimpan user ke context
			c.Set("user", user)
			return next(c)
		}
	}
}
